File: LeetCode/71. Simplify Path.py
# https://leetcode.com/problems/simplify-path/description/
from test import test


class Solution:
    UP = ".."
    CURRENT = "."

    def simplifyPath(self, path: str) -> str:
        array = path.split("/")

        print(array)

        result: list[str] = []

        for segment in array:
            if segment == self.UP and result:
                result.pop()
            elif segment in [self.CURRENT, self.UP, ""]:
                continue
            else:
                result.append(segment)

        return "/" + "/".join(result)


solution = Solution()


test(
    solution.simplifyPath,
    [
        {
            "input": ["/home/"],
            "expected": "/home",
        },
        {
            "input": ["/home//foo/"],
            "expected": "/home/foo",
        },
        {
            "input": ["/home/user/Documents/../Pictures"],
            "expected": "/home/user/Pictures",
        },
        {
            "input": ["/../"],
            "expected": "/",
        },
        {
            "input": ["/.../a/../b/c/../d/./"],
            "expected": "/.../b/d",
        },
    ],
)

File: LeetCode/71. Simplify Path.js
// https://leetcode.com/problems/simplify-path/description/

const SKIP = '.';
const UP = '..';

const simplifyPath = (path) => {
  const formattedPath = path.replace(/\/+/g, '/');

  const arr = formattedPath.split('/').filter(Boolean);

  const result = [];

  for (const segment of arr) {
    if (segment === SKIP) {
      continue;
    }

    if (segment === UP) {
      result.pop();
      continue;
    }

    result.push(segment);
  }

  return '/' + result.join('/');
};

console.log(simplifyPath('/home/')); // /home
console.log(simplifyPath('/home//foo/')); // /home/foo
console.log(simplifyPath('/home/user/Documents/../Pictures')); // /home/user/Pictures
console.log(simplifyPath('/../')); // /
console.log(simplifyPath('/.../a/../b/c/../d/./')); // /.../b/d
